fraud/bad_encoding_test.go

```diff
@@ -0,0 +1,180 @@
+package fraud
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/pkg/da"
+
+	"github.com/celestiaorg/celestia-node/ipld"
+	"github.com/celestiaorg/celestia-node/service/header"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+func TestCreateBadEncodingFraudProof(t *testing.T) {
+	eds := ipld.GenerateRandEDS(t, 2)
+	type test struct {
+		name   string
+		isRow  bool
+		roots  func(uint) [][]byte
+		length int
+	}
+	tests := []test{
+		{"rowRoots", true, eds.Row, len(eds.RowRoots())},
+		{"colRoots", false, eds.Col, len(eds.ColRoots())},
+	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			for index := 0; index < tc.length; index++ {
+				_, err := CreateBadEncodingFraudProof(1, uint8(index), tc.isRow, eds, tc.roots(uint(index)))
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestFraudProofValidationForRow(t *testing.T) {
+	eds := ipld.GenerateRandEDS(t, 2)
+	size := eds.Width()
+
+	tree := NewErasuredNamespacedMerkleTree(uint64(size / 2))
+	shares := flatten(eds)
+
+	copy(shares[0][8:], shares[1][8:])
+
+	eds1, err := rsmt2d.ImportExtendedDataSquare(shares, rsmt2d.NewRSGF8Codec(), tree.Constructor)
+	require.NoError(t, err)
+	dataSquare := make([][]byte, size*size)
+	copy(dataSquare, shares)
+	dataSquare[2] = nil
+	dataSquare[3] = nil
+	dataSquare[8] = nil
+	dataSquare[12] = nil
+	_, err = rsmt2d.RepairExtendedDataSquare(
+		eds1.RowRoots(),
+		eds1.ColRoots(),
+		dataSquare,
+		rsmt2d.NewRSGF8Codec(),
+		tree.Constructor,
+	)
+
+	require.Error(t, err)
+	// [2] and [3] will be empty
+	var errRow *rsmt2d.ErrByzantineRow
+	require.True(t, errors.As(err, &errRow))
+
+	errRow = err.(*rsmt2d.ErrByzantineRow)
+	fmt.Println(errRow.Shares)
+
+	p, err := CreateBadEncodingFraudProof(1, uint8(errRow.RowNumber), true, eds1, errRow.Shares)
```

Comment:
Shouldn't we use here the eds returned from repair on loc 57?

Reply:
I also think that was the conclusion of a call with @adlerjohn one or two weeks ago 🤔

Reply:
No, because repairing process can return new eds which is not the same as `eds1`. As Markle proof validation should be done against our bad eds, we should you it for creation also.